Update facilities tableDevelop update Facilities DB from reviewed geopackage

pkinglinz opened the pull request:
Upload reviewed gpkg to the temp_facilities table in the database
Update facilities table
- update attributes on “facilities” table from temp upload table, based on “sql”
- update geometries on “facilities” table from temp upload table, based on “geometry_change” and feature geometry

copy of symlink the update_facilities folder of the repo to your plugin directory in qgis, I used ln -s ~/dev/nz-facilities/update_facilities/ ~/.local/share/QGIS/QGIS3/profiles/default/python/plugins

create the config
the plugin connects to db via a config file. This file is autogenerated if it doesn't exist but user will need to find and update the link to the database. 
![image](https://github.com/user-attachments/assets/8e306ff5-ab1a-400e-88d6-17e3e950252d)
then navigate to the update_facilities folder and select the config.ini file

I suggest you create a test db to connect to, this is the dump file I was using and the command used on ubuntu to restore it
https://linzsrm.sharepoint.com/:u:/s/Topography/ES9g9Rysll5IpQEPoyBG2Q0BXP3iYNAu8_quettsfsjoXg?e=eI2N8l

pg_restore -h localhost -p 5432 -d facilities_test_db -U postgres -O -v -x --no-table-access-method "/home/pking/dev/nz-facilities/change_detection/facilities_dump/facilities.backup"

you will need to add the logging tables and update the temp_facilities table, there is an sql file with some quick code to do this from SharePoint, https://linzsrm.sharepoint.com/:f:/s/Topography/IgBWkmu7txi-QL_UnlByArmYAQfT0Wtw26UlGgALn1b7Y-c?e=3CAV3c

I have also put test data and an output from the facilities script in SharePoint

pkinglinz committed:
fix: schools layer name

fixed name of MoE schools layer on data.govt. Assume it is changed everytime they update the layer

pkinglinz committed:
feat: upload facilities to temp table

set up plugin, my file is update_facilities.py. tes_dbconn.py test the database connection, update_temp_facilities.py handles updating the temp_facilities table in the database. A config will be created with correct sections but will need to be manually configured

pkinglinz committed:
feat: process to update facilities table

Updates facilities table using the temp table and attribute in the change_action column. If errors are encountered it is rolled back.

pkinglinz committed:
feat: test before run

Test before running updates. Move run to class. Text red on fail, green on pass.

pkinglinz committed:
feat: add task logging

Logging via DB so anyone from team can access. Logs what task it is doing, when starts task, errors and final result of task.

pkinglinz committed:
feat: log results of updating facilities table

logs single line of added, delete, modified geom, mod attr and modified both geom and attr

pkinglinz committed:
fix: logging

Now only logs results when successful. Added type hints to logging function

pkinglinz committed:
Style: add type hints

Douglas-K reviewed (changes requested):
This works very well.  Nice easy plugin to use: config setup, simple UI, correctly updates temp_facilities and facilities tables in the DB, and good UI reporting and DB logging.
Here are some adjustments I recommend:

- UI reporting:
![image](https://github.com/user-attachments/assets/aaba75dc-3cfa-4a24-aaf3-7c9fc93a805b)
![image](https://github.com/user-attachments/assets/c2b08103-a5ea-4637-bf71-d4f0a8b5b339)
-- Remove extra blank lines
-- For failed updates, either remove count of features changed, or reword and reorder it so that it shows first as count of features it is trying to change.
-- For failed features, use one line for each, plus reason, keep it brief.  eg:
    --- Facility ID 308: Failed to delete. Could not find
    --- Facility ID 479: Failed to modify geom. New matches old
    --- Facility ID xxx: Failed to add. ID already exists
    --- Facility ID xxx: Failed to modify attr. New matches old
-- Add fail tests for adding a geom when the ID already exists, and modifying attributes when there is no change.
-- Add fail test for adding / modifying geom causing an intersection error with other geoms.  (Or maybe just check all geoms for intersections?) eg:
    --- Facility ID xxx: Failed overlap check.

- temp_facilities table:
![image](https://github.com/user-attachments/assets/570e0e61-2826-4c4f-be37-3e5286c9360d)
-- Clear out error_description column at the start of re-running "Update Facilities Table", as it is currently appending.
-- Simplify error_description text to match the updated UI messages (above), excluding text "Facility ID xxx: "

- Result logging:
![image](https://github.com/user-attachments/assets/963a7732-55f7-452a-b0f9-ab240c022b0f)
-- Add a count for before update
-- Add a count for unchanged features
-- Add a count for after update

pkinglinz committed:
feat: update facilities

Workflow and UI to update the lds_facilities table

pkinglinz committed:
fix: remove extra line spacing from ui messages

pkinglinz committed:
fix: report unchanges

Change how report when no changes so clear unchanged, and report expected changes. Adjusted sql file so alphabetical and added in query.

pkinglinz committed:
feat: check for duplicates, change reporting

looks to see if name, source_name, source_facility_id or geometry match any other features. reporting to viewer simplified

pkinglinz committed:
fix: facilities row count text

Made more readable so states exactly what row count refering to

pkinglinz committed:
feat: check for overlapping polygons

Finds when the added or modified geometry overlaps other geometries

pkinglinz committed:
fix: clear error description and simplify message

No longer appends error description to the tmp_facilities table, clears - sets to null, all error descriptions for attempting to update the facilities table. Uses the simplified message used in the message box. No need to include sql as this found in table if tis modifying, or it is adding or deleting.

pkinglinz committed:
feat: add additional results log

Add before and after row count, and number unchanged. Also adjusted the overlap so snaps to grid to avoid slight overlaps

pkinglinz committed:
feat: consistent logs

ui message and log messages now have consistent format

pkinglinz commented:
note the logging tables need to be updated as have additional columns. The sql to update these can be found in the create_tables.sql file, the two tables are:
facilities.facilities_task_logging
facilities.facilities_result_logging
adding note here as failed for reviewer, process will run but then will be message in the message box of plugin about update columns not correct

Douglas-K commented:
This works well.  Thanks for the changes and other enhancements.

Note that update_lds_facilities_table.py runs checks but doesn't seem to update the lds facilities table yet.  However, we will likely do this another way, by deleting the lds facilities table and copying the production facilities table over as a fresh copy (similar to how Building Outlines works).  We will deal with this separately.
